iex/json_serializer.h

```diff
@@ -0,0 +1,47 @@
+/**
+ * Copyright 2020 Antony Kellermann
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+ * documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+ * Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+ * WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#pragma once
+
+#include <nlohmann/json.hpp>
+
+namespace iex::json
+{
+using Json = nlohmann::json;
+
+/**
+ * This is a base class for objects that can serialized and deserialized with Json data.
+ *
+ * As this is an abstract interface, all functions must be overridden.
+ */
+class JsonSerializable
+{
+ protected:
+  /**
+   * Creates and returns a Json object representing this object. This function must be overridden.
+   * @return Json
+   */
+  virtual ValueWithErrorCode<Json> Serialize() = 0;
```

Comment:
Serialize and Deserialize should probably be put in separate interfaces. Not all derived classes will require both functions, such as the one in #37.